### PR TITLE
Override Vector operators in Point2 and Point3

### PR DESCRIPTION
**Pull request by jeffin07**
## Motivation and Context

Overrides vector operations for `Point2`, `Point3` #113 

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by jeffin07** [comment]
@luisenp overrides `add`, `sub`, `mul`, `truediv` Please review.Will add test after finalizing

**Review by luisenp — CHANGES_REQUESTED** [review]
This looks good me, just need to add unit tests :)

**Review by mhmukadam — COMMENTED** [review]
Looks good. Luis mentioned unit tests already.

Other functions in Vector that return Vector and need overriding (@luisenp please confirm): neg, abs, cat, _retract_impl, _compose_mpl, _inverse_impl

**Comment by luisenp** [comment]
> @luisenp we're already checking for the return type [here](https://github.com/facebookresearch/theseus/blob/main/theseus/geometry/tests/test_point_types.py#L32l)
> 
> Do we need any additional unit test for the return type ?

Oh, right, we have some test that check that the dynamic type is correct. However, this PR is mostly for mypy type checking, so that test won't capture this behavior. 

Here is one way to test this. You can add the following test:

```python
def test_operators_mypy_cast():
    import mypy.api.  # Put this import at the top of the file

    result = mypy.api.run(["theseus/geometry/tests/point_types_mypy_check.py"])
    assert result[2] == 0
```

And then in file `theseus/geometry/tests/point_types_mypy_check.py` do something like the following:

```python
import theseus as th

x = th.Point2()
y = th.Point2()
z: th.Point2 = x + y
``` 

This test fails for me and it shouldn't fail after the changes in this PR.

**Comment by jeffin07** [comment]
@luisenp  added test please review

**Comment by luisenp** [comment]
LGTM! Let's now add `cat`, `neg`, `abs`, and `exp_map`. The functions with `_impl_` in the name are not user facing, so we don't need to fix the casting for those.

**Comment by jeffin07** [comment]
@luisenp `exp_map` is already returns correct type do we need to change it ?. Also can you please help me about how to test `cat` method

**Comment by jeffin07** [comment]
@luisenp  I have added the changes. Please review

**Comment by mhmukadam** [comment]
LGTM. The CI is failing because mypy isn't available, might need to update the CI config.

**Comment by luisenp** [comment]
@jeffin07 Can you try adding `pip install -r requirements/dev.txt` below [this](https://github.com/facebookresearch/theseus/blob/main/.circleci/config.yml#L112) line?

**Comment by jeffin07** [comment]
@luisenp Added the CI changes

**Review by mhmukadam — APPROVED** [review]

File: theseus/geometry/tests/point_types_mypy_check.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import theseus as th
+
+
+def mypy_operations_check():
```
Comment (luisenp) [inline theseus/geometry/tests/point_types_mypy_check.py:4]: For some reason `mypy` doesn't catch any errors if this function definition is here. We don't need this, so let's remove it.

File: theseus/geometry/tests/point_types_mypy_check.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+    for point_cls in [th.Point2, th.Point3]:
```
Comment (luisenp) [inline theseus/geometry/tests/point_types_mypy_check.py:6]: I tested this locally, and `mypy` doesn't like the syntax `z: point_cls`. I get this error:

```bash
error: Variable "theseus.geometry.tests.point_types_mypy_check.point_cls" is not valid as a type  [valid-type]
theseus/geometry/tests/point_types_mypy_check.py:9: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

I suggest to remove the loop and replace all `: point_cls` for with `: th.Point2` (and again for `: th.Point3`). Also, `mypy` doesn't like repeating the variable `z`, so let's use `z1`, `z2`, `z3`, etc.

**Review by luisenp — CHANGES_REQUESTED** [review]
Need to fix the unit tests, but otherwise almost done. Thanks!

**Comment by jeffin07** [comment]
@luisenp  Added the changes but test for cuda fails

**Comment by luisenp** [comment]
@jeffin07 There seems to be some issue affecting all of our PRs right now, I'm looking into it. Your latest changes look good!

**Review by luisenp — APPROVED** [review]

**Comment by luisenp** [comment]
@jeffin07 This looks ready to me, thanks again for your contribution!

**Pull request merged**